mavros/src/plugins/setpoint_global_position.cpp

```diff
@@ -0,0 +1,192 @@
+/**
+ * @brief SetpointGlobalPosition plugin
+ * @file setpoint_global_position.cpp
+ * @author Mohamed Abdelkader 2017 <mohamedashraf123@gmail.com>
+ *
+ * @addtogroup plugin
+ * @{
+ */
+/*
+ * Copyright 2017, Mohamed Abdelkader.
+ *
+ * This file is part of the mavros package and subject to the license terms
+ * in the top-level LICENSE file of the mavros repository.
+ * https://github.com/mavlink/mavros/tree/master/LICENSE.md
+ */
+
+#include <mavros/mavros_plugin.h>
+#include <mavros/setpoint_mixin.h>
+#include <eigen_conversions/eigen_msg.h>
+
+#include <geometry_msgs/PoseStamped.h>
+#include <sensor_msgs/NavSatFix.h>
+#include <mavros_msgs/GlobalPositionTarget.h>
+
+#define M_DEG_TO_RAD 0.01745329251994
+#define CONSTANTS_RADIUS_OF_EARTH 6371000.0
+//#define DBL_EPSILON 2.2204460492503131E-16
+
+namespace mavros {
+namespace std_plugins {
+/**
+ * @brief Setpoint global position plugin
+ *
+ * Send global setpoint positions to FCU controller.
+ */
+class SetpointGlobalPositionPlugin : public plugin::PluginBase,
+	private plugin::SetPositionTargetLocalNEDMixin<SetpointGlobalPositionPlugin>{
+public:
+	SetpointGlobalPositionPlugin() : PluginBase(),
+		// NOTE not private handle, because we need to access gps topic, which is not under this node. Is it OK?
+		sp_nh("~")
+	{ }
+
+	void initialize(UAS &uas_)
+	{
+		PluginBase::initialize(uas_);
+
+		/* subscriber for current gps state, mavros/global_position/global. */
+		gps_sub = sp_nh.subscribe("global_position/global", 10, &SetpointGlobalPositionPlugin::gps_cb, this);
+		/* Subscribe for current local pose. */
+		local_sub = sp_nh.subscribe("local_position/pose", 10, &SetpointGlobalPositionPlugin::local_cb, this) ;
+		/* Subscriber for target gps state */
+		setpoint_sub = sp_nh.subscribe("setpoint_global_position/global", 10, &SetpointGlobalPositionPlugin::setpoint_cb, this);
+
+	}
+
+	Subscriptions get_subscriptions()
+	{
+		return { /* Rx disabled */ };
+	}
+
+private:
+	friend class SetPositionTargetLocalNEDMixin;
+	ros::NodeHandle sp_nh;
+
+	ros::Subscriber setpoint_sub;
+	ros::Subscriber gps_sub;
+	ros::Subscriber local_sub;
+
+	/* resulting change in x/y in NED frame. Set by lla2ned() function. */
+	double dx, dy;
+
+	/* Stores current gps state. */
+	sensor_msgs::NavSatFix gps_msg;
+	/* Current local pose */
+	geometry_msgs::PoseStamped local_pose_msg;
+	/* old time gps time stamp in [ms], to check if new gps msg is received */
+	uint32_t old_gps_stamp = 0;
+
+	/* -*- mid-level helpers -*- */
+
+	/**
+	 * @brief Converts lat/long to 'change' in local NED frme
+	 * @param curr_lat. Current latitude in deg
+	 * @param curr_long. Current longitude in deg
+	 * @param target_lat. Target latitude in deg
+	 * @param target_long. Target longitude in deg
+	 */
+	 void lla2ned(double curr_lat, double curr_long, double target_lat, double target_long){
```

Comment:
Why a new function like this? Did you check how `fake_gps` handles the conversion from geodetic to NED coordinates? It is everything there. You don't need to create a new function like this while you can use GeographicLib functions to do that job.